src/core/mac/sub_mac.cpp

```diff
@@ -710,7 +718,12 @@ bool SubMac::ShouldHandleCsmaBackOff(void) const
 {
     bool swCsma = true;
 
+#if OPENTHREAD_CONFIG_MULTIPAN_RCP_ENABLE
+    // Use software csma to perform retries in case of a busy RCP
+    goto exit;
+#else
     VerifyOrExit(!RadioSupportsCsmaBackoff(), swCsma = false);
+#endif
```

Comment:
Hi Abtin, I agree, I did not like those changes to the OT core modules. This change is for the case where one process wants to tx while the other one is already transmitting.  I can place logic to handle it (basically performing retries) into radio_spinel_impl.hpp.  Does that seem like an ok place for it?

Thanks for the additional discussion, it is very helpful.  In your case were your stack entities running in a single process so that the arbiter module sat under both of them, or were they in separate processes?  If the latter, how did the IPC work?

Reply:
Thanks @mspang.

I agree that handling it in `radio_spinel_imp.hpp` sounds reasonable. I would say if we can move it to RCP side itself, it may be even better. I think RCP  is the place where we need to be aware of multiple PANs/entities and it can best handle any conflicting requests.

I think from the host side of `radio_spinel_impl`, we may only need to know to use certain spinel IID to talk to the RCP (would it matter to know if there are other entities that are controlling the RCP or not on the host side?).

----------

Let me try add more details (from what I recall) on our internal project and how it worked:

- It was all running on an RTOS platform. All modules were part of the firmware running on NPC (OT stack, radio arbiter and driver, and other (legacy) stacks) all part of NCP firmware. 
- OT stack had its own RTOS task:
   - OT stack requires that all the OT public and platform APIs and callbacks to be invoked from same OS context, e.g. same RTOS task.
- Arbiter and radio driver had their own shared RTOS task (and I think other/legacy stack ran on its own RTOS task as well).
- I think the different tasks synced using ROTS locks/semaphore.

Arbiter module provided APIs to all entities which wanted to control the radio:
- For example, `otPlatRadio` APIs to OT stack and very similar set of APIs to the other (legacy) stack.
- Arbiter module managed state variables per entity to remember the last request/command (API call). 
   - For example, if `otPlatRadioSleep()` was called to put radio to sleep, or `otPlatRadioReceive()` to rx, or a request to tx frame or perform energy scan, etc.
- It then handled the requests/commands from different entities and decided what to do actually with the radio. Here are some of the rules I recall:
   - We put radio to sleep mode only when all entities requested it to sleep.
   - When different entities requested radio to be in rx mode: 
        - If the channels matched, it was straight-forward and we put the radio in rx mode on that channel.
        - But if the requests were for different channels, I think we had a simple rule to prioritize one entity (I think OT stack request won on the rx channel).
        - Just a note on this: 
            - In our use-case, all stacks (OT and legacy) were actually operating on the same channel, but we still could have the situation that they put radio on different channels, for example to support a discover/mac scan operation (during which we can switch to different channels).
            - We also had the idea to use a windowed rx model when we have conflicting rx channels (rx on first channel for some time interval, then switch to other channel, an so on) but I think we did not implement/use this model.
   - If we get a tx request by an entity while radio was busy with an ongoing tx from another entity we just waited for the current tx to complete (success or failure) and then proceeded with the next tx request.

Reply:
Hi @abtink, thanks for this discussion.  I would like to move the arbitration of transmissions and energy scans to the RCP side as you suggest.  I can add a queue to NcpBase which queues up transmission and energy scan commands if the radio is busy, then proceeds with them when the radio is free. That code would only be enabled in MULTIPAN_RCP mode.  Let me know if that sounds ok.

Reply:
Thanks @mnp222 . Yes, I think it makes sense and sounds good to me. 👍 

I think we can also consider adding this in the `LinkRaw` class (under `OPENTHREAD_RADIO`) since `LinkRaw` is the point of interaction with radio in an RCP build. `LinkRaw` practically provides a mirror of `radio.hpp` APIs.  

I think placing the multi-pan (radio busy) checks in either `Ncp` or `LinkRaw` would be fine. Thanks.